downloadFileContent(): Follow URL redirects in Node [#585]

HenrikBengtsson opened the pull request.

HenrikBengtsson committed:
downloadFileContent(): Follow URL redirects in Node [#585]

dirkschumacher commented on src/webR/webr-worker.ts:
```diff
@@ -610,15 +614,38 @@ function downloadFileContent(URL: string, headers: Array<string> = []): XHRRespo
+
+      if (location) {
+        // Resolve relative URLs against the original URL
+        const redirectUrl = new URL(location, url).href;
```
`new URL` can throw a `TypeError` exception in case the url is not valid. That should be caught here IMHO.

HenrikBengtsson replied:
Great point. I've updated to handler any errors - including the special case TypeError. Can you please take a look?

dirkschumacher commented on src/webR/webr-worker.ts:
```diff
@@ -592,9 +592,13 @@ function copyFSNode(obj: FSNode): FSNode {
 
-function downloadFileContent(URL: string, headers: Array<string> = []): XHRResponse {
+function downloadFileContent(url: string, headers: Array<string> = [], maxRedirects: number = 10): XHRResponse {
+  if (maxRedirects == 0) {
```
Maybe change to `maxRedirects <= 0` - just in case to prevent future potential bugs. Right now it is correct though

HenrikBengtsson replied:
Thanks. Added.

dirkschumacher commented on src/webR/webr-worker.ts:
```diff
@@ -610,15 +614,38 @@ function downloadFileContent(URL: string, headers: Array<string> = []): XHRRespo
+
+      if (location) {
+        // Resolve relative URLs against the original URL
+        const redirectUrl = new URL(location, url).href;
```
The URL can be "be relative to the request URL or an absolute URL" https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Location#url

The case of an absolute URL probably creates an exception here.

HenrikBengtsson replied:
> The case of an absolute URL probably creates an exception here.

The example of #585 uses an absolute URL redirect.

According to https://developer.mozilla.org/en-US/docs/Web/API/URL/URL, it sounds like `url` is ignored if `location` is an absolute URL;

> A string or any other object with a stringifier that represents an absolute URL or a relative reference to a base URL. If `url` is a relative reference, `base` is required, and is used to resolve the final URL. If `url` is an absolute URL, a given `base` will not be used to create the resulting URL.

Was that your concern here?

PS. I'm not a regular JavaScript developer.

dirkschumacher replied:
Correct, in that case it's seems to be fine

HenrikBengtsson committed:
downloadFileContent(): handle maxRedirects < 0 (just in case)

HenrikBengtsson committed:
downloadFileContent(): handle redirect URL errors

HenrikBengtsson commented on src/webR/webr-worker.ts:
```diff
@@ -610,15 +614,50 @@ function downloadFileContent(URL: string, headers: Array<string> = []): XHRRespo
+      status = request.status;
+    }
+
+    // Follow 3xx redirects
```
The browser already takes care of redirects. Should the code for handling 3xx redirects be conditioned by `IN_NODE`, e.g.

```js
    if (IN_NODE && status >= 300 && status < 400) {
       ...
    }
 ```

?

HenrikBengtsson replied:
If this is done, the code can be cleaned up even further.

HenrikBengtsson replied:
I've consolidated all `IN_NODE` code to one `if (IN_NODE) { ... }` statement.

HenrikBengtsson committed:
downloadFileContent(): consolidate IN_NODE-specific code

HenrikBengtsson committed:
Address @typescript-eslint/no-inferrable-types;

Error: 595:72 error Type number trivially inferred from a number literal, remove type annotation @typescript-eslint/no-inferrable-types

HenrikBengtsson commented:
The GA build-webr checks produced:

```
/__w/webr/webr/src/webR/webr-worker.ts
Error:   595:72  error  Type number trivially inferred from a number literal, remove type annotation  @typescript-eslint/no-inferrable-types
```

I address that in commit 1f2011e. @georgestagg , could please trigger a rerun?

georgestagg commented:
Yeah, the linter is configured to be quite picky…

You may be able to check it locally quicker by cd-ing into `src` and running `make lint`

georgestagg reviewed (approved):
LGTM!

If we wanted, we could add a test for `download.file()` in the same place as we already have a couple of tests that depend on networking [here](https://github.com/r-wasm/webr/blob/71fa42f80dbb7e72602df180e9500c269ab70134/src/tests/webR/webr-worker.test.ts#L40).

Happy to merge without, but it's an option if you'd like to add it. We'd need to test downloading a file from with a reasonable URL that will always be consistent and available.

HenrikBengtsson committed:
Add tests for installing package from R-universe, which involves URL redirect

HenrikBengtsson committed:
Add tests for utils::download.file() with and without URL redirects
